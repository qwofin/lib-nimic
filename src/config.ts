import pino from 'pino'

class Config {
  downloadDir: string
  completedDir: string
  apiPort: number
  nntpPort: number
  chunkSize: number
  logger: pino.Logger
  nzbHostOverride?: string
  pinoOptions: pino.LoggerOptions
  libgenMirror: { url: string; dlUrl: string }
  libgenPlusMirror: { url: string }
  rateLimit: {
    rate: number
    timeframe: number
  }
  ipfsGw: string

  constructor() {
    this.rateLimit = {
      rate: 40,
      timeframe: 300,
    }

    this.downloadDir = process.env.DOWNLOAD_DIR
    this.completedDir = process.env.COMPLETED_DIR
    if (!this.completedDir) throw new Error('COMPLETED_DIR must be set')
    if (!this.downloadDir) throw new Error('DOWNLOAD_DIR must be set')
    this.apiPort = parseInt(process.env.API_PORT ?? '3000')
    this.chunkSize = parseInt(process.env.CHUNK_SIZE ?? String(1024 * 1024 * 3))
    this.pinoOptions = {
      level: process.env.LOG_LEVEL ?? 'info',
      enabled: process.env.NODE_ENV !== 'test',
      mixin: (_context, level) => {
        return { severity: pino.levels.labels[level] }
      },
    }

    if (process.env.LOG_FORMAT === 'pretty') {
      this.pinoOptions.transport = {
        target: 'pino-pretty',
        options: {
          hideObject: true,
          ignore: 'pid,hostname',
          messageFormat: '{module}| {msg}',
          translateTime: 'yyyy-mm-dd HH:MM:ss',
        },
      }
      if (process.env.LOG_COLORS === 'false') {
        this.pinoOptions.transport.options.colorize = false
      }
      if (process.env.LOG_SHOW_OBJECTS === 'true') {
        this.pinoOptions.transport.options.hideObject = false
      }
    }
    this.logger = pino.pino(this.pinoOptions).child({ module: 'main' })
    this.libgenMirror = {
      url: (process.env.LIBGEN_MIRROR ?? '').replace('/*$', '/'),
      dlUrl: (process.env.LIBGEN_MIRROR_FILEHOST ?? '').replace('/*$', '/'),
    }
    this.libgenPlusMirror = {
      url: (process.env.LIBGENPLUS_MIRROR ?? '').replace('/*$', '/'),
    }

    if (this.libgenMirror.url && !this.libgenMirror.dlUrl)
      throw new Error(
        'When LIBGEN_MIRROR is set LIBGEN_MIRROR_FILEHOST must also be set',
      )
    if (!this.libgenMirror.url && !this.libgenPlusMirror.url)
      throw new Error('No sources configured')

    this.ipfsGw = (process.env.IPFS_GW ?? '').replace('/*$', '/')
  }
}
export const config: Config = new Config()
