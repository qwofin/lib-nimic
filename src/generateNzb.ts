import Path from 'path'
import xml2js from 'xml2js'
const META_PREFIX = 'x-lib-nimic'

export interface ParsedNzb {
  dlurl: string
  md5: string
  filename: string
  size: number
  sourcetype: string
}

export function generateNzb(
  dlurl: string,
  filename: string,
  size: number,
  md5: string,
  sourcetype: string,
): [string, string] {
  const name = Path.parse(filename).name
  const xAttribs: { [key: string]: string } = {
    dlurl,
    md5,
    filename,
    size: size.toString(),
    sourcetype,
  }

  const builder = new xml2js.Builder({
    doctype: {
      pubID: '-//newzBin//DTD NZB 1.1//EN',
      sysID: 'http://www.newzbin.com/DTD/nzb/nzb-1.1.dtd',
    },
  })
  const body = builder.buildObject({
    nzb: {
      $: { xmlns: 'http://www.newzbin.com/DTD/2003/nzb' },
      head: {
        meta: [
          { $: { type: 'category' }, _: 'Books > Ebooks' },
          { $: { type: 'name' }, _: name },
          ...Object.entries(xAttribs).map(([key, value]) => {
            return { $: { type: `${META_PREFIX}:${key}` }, _: value }
          }),
        ],
      },
      file: {
        $: { subject: filename },
        groups: { group: 'alt.binaries.e-book' },
        segments: {
          segment: {
            $: { bytes: size, number: 1 },
            _: `1@lib-nimic`,
          },
        },
      },
    },
  })

  return [`${filename}.nzb`, body]
}

export async function parseNzb(nzb: string): Promise<ParsedNzb> {
  const parser = new xml2js.Parser()
  const parsed = await parser.parseStringPromise(nzb)
  const meta = parsed.nzb.head[0].meta

  return {
    dlurl: meta.find(item => item['$'].type == `${META_PREFIX}:dlurl`)['_'],
    md5: meta.find(item => item['$'].type == `${META_PREFIX}:md5`)['_'],
    filename: meta.find(item => item['$'].type == `${META_PREFIX}:filename`)[
      '_'
    ],
    size: parseInt(
      meta.find(item => item['$'].type == `${META_PREFIX}:size`)['_'],
    ),
    sourcetype: meta.find(
      item => item['$'].type == `${META_PREFIX}:sourcetype`,
    )['_'],
  }
}
