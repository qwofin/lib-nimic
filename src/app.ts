import Koa from 'koa'
// import logger from 'koa-pino-logger'
import { pinoHttp } from 'pino-http'
import { router } from './api/routes.js'
import { config } from './config.js'
const app = new Koa()
const logger = config.logger.child({ module: 'api' })
const logWrapper = pinoHttp({ logger })
// const apiLogger = logger(config.pinoOptions)
// apiLogger.logger.setBindings({ module: 'api' })

app.use(async (ctx, next) => {
  try {
    await next()
  } catch (err) {
    ctx.status = err.statusCode || err.status || 500
    ctx.body = {
      error: err.message,
    }
  }
})
app.use(async (ctx, next) => {
  logWrapper(ctx.req, ctx.res)
  try {
    await next()
  } catch (err) {
    ctx.req.log.error({ err })
    throw err
  }
})
app.use(router.routes())
export default app
