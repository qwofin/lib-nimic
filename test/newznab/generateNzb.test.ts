import { config } from '../../src/config.js'
import request from 'supertest'
import app from '../../src/app.js'
import { parseNzb } from '../../src/generateNzb.js'

describe('nzb generation', () => {
  test('generates valid nzb', async () => {
    config.chunkSize = 1024
    const response = await request(app.callback()).get(
      `/api/nzb?dlurl=https://test.com/test.epub&size=3073&md5=abcdef&filename=test.epub&sourcetype=libgen`,
    )
    expect(response.status).toBe(200)
    expect(response.text).toMatchSnapshot()
  })

  test('parses valid nzb', async () => {
    expect(
      await parseNzb(
        (
          await request(app.callback()).get(
            `/api/nzb?dlurl=https://test.com/test.epub&size=3073&md5=abcdef&filename=test.epub&sourcetype=libgen`,
          )
        ).text,
      ),
    ).toEqual({
      dlurl: 'https://test.com/test.epub',
      size: 3073,
      md5: 'abcdef',
      filename: 'test.epub',
      sourcetype: 'libgen',
    })
    expect(
      await parseNzb(
        (
          await request(app.callback()).get(
            `/api/nzb?dlurl=https://test.com/test.epub&size=3073&md5=abcdef&filename=test.epub&sourcetype=libgen`,
          )
        ).text,
      ),
    ).toEqual({
      dlurl: 'https://test.com/test.epub',
      size: 3073,
      md5: 'abcdef',
      filename: 'test.epub',
      sourcetype: 'libgen',
    })
  })
})
