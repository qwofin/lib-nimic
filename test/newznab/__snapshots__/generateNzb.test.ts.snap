// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nzb generation generates valid nzb 1`] = `
"<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<!DOCTYPE nzb PUBLIC "-//newzBin//DTD NZB 1.1//EN" "http://www.newzbin.com/DTD/nzb/nzb-1.1.dtd">
<nzb xmlns="http://www.newzbin.com/DTD/2003/nzb">
  <head>
    <meta type="category">Books &gt; Ebooks</meta>
    <meta type="name">test</meta>
    <meta type="x-lib-nimic:dlurl">https://test.com/test.epub</meta>
    <meta type="x-lib-nimic:md5">abcdef</meta>
    <meta type="x-lib-nimic:filename">test.epub</meta>
    <meta type="x-lib-nimic:size">3073</meta>
    <meta type="x-lib-nimic:sourcetype">libgen</meta>
  </head>
  <file subject="test.epub">
    <groups>
      <group>alt.binaries.e-book</group>
    </groups>
    <segments>
      <segment bytes="3073" number="1">1@lib-nimic</segment>
    </segments>
  </file>
</nzb>"
`;
